### PR TITLE
[Feature templates] Device model validation

### PR DESCRIPTION
**Pull request by cicharka**
# Pull Request summary:
Functionality to protect users from creating and attaching templates that are not valid for provided devices.

# Description of changes:
In order to validate device models available for template, endpoint `"/dataservice/template/feature/types"` is used to fetch all of the template types and associated device models. Then, by forcing user to provide list of device models when preparing the template, the same list is compared with available device models for desired template.

# Checklist:
- [x] Make sure to run pre-commit before commiting changes
- [x] Make sure all checks have passed
- [x] PR description is clear and comprehensive
- [x] Mentioned the issue that this PR solves (if applicable)
- [x] Make sure you test the changes

Results:
1. When providing proper devices, validation pass without notifying user
2. If user provides device model in list of device models that is not supported for the template, `DeviceModelError` is raised, and debug message is added to log with information which devices are supported.

### CODE REVIEW

File: vmngclient/dataclasses.py [hunk 1]
```diff
@@ -475,3 +475,17 @@ class TenantTacacsServer(DataclassBase):
...
+@define
+class FeatureTemplatesTypes(DataclassBase):
+    parent: str = field(metadata={FIELD_NAME: "parent"})
+    default: str = field(metadata={FIELD_NAME: "default"})
+    display_name: str = field(metadata={FIELD_NAME: "displayName"})
+    name: str = field(metadata={FIELD_NAME: "name"})
+    type_class: str = field(metadata={FIELD_NAME: "typeClass"})
+    description: str = field(metadata={FIELD_NAME: "description"})
+    write_permission: str = field(metadata={FIELD_NAME: "write_permission"})
+    read_permission: str = field(metadata={FIELD_NAME: "read_permission"})
```
Comment (tomaszwarda) [inline vmngclient/dataclasses.py:489]: Isn't `FIELD_NAME` only required when it is different than name?

Comment (cicharka) [reply vmngclient/dataclasses.py:489]: True, will fix

**Review by tomaszwarda — DISMISSED** [review]

**Review by JimOverholt — DISMISSED** [review]

File: vmngclient/api/templates/feature_template.py [hunk 2]
```diff
@@ -2,18 +2,21 @@
...
+from vmngclient.utils.device_model import DeviceModel
+
 if TYPE_CHECKING:
     from vmngclient.session import vManageSession
 
 
 class FeatureTemplate(BaseModel, ABC):
     name: str
     description: str
+    device_models: List[Union[DeviceModel, str]]
```
Comment (tehAgitto) [inline vmngclient/api/templates/feature_template.py:19]: I think that we can use only `DeviceModel` object. Pydantic should convert strings into our objects instantly.

File: vmngclient/api/template_api.py [hunk 3]
```diff
@@ -531,6 +540,27 @@ def generate_feature_template_payload(
...
+        provided_device_models = [
+            dev_mod.value if type(dev_mod) is DeviceModel else dev_mod for dev_mod in template.device_models
+        ]
+
+        if not all(dev in available_devices_for_template for dev in provided_device_models):
+            logger.debug(f"Available devices for template '{template.type}': {available_devices_for_template}")
+            raise DeviceModelError(template, provided_device_models)
+        return True
+
+    def _get_feature_template_types(self, type: str = "all"):
```
Comment (tehAgitto) [inline vmngclient/api/template_api.py:557]: We're missing return type.

File: vmngclient/dataclasses.py [hunk 4]
```diff
@@ -477,3 +477,17 @@ class TenantTacacsServer(DataclassBase):
...
+    parent: str
+    default: str
+    display_name: str = field(metadata={FIELD_NAME: "displayName"})
+    name: str
+    type_class: str = field(metadata={FIELD_NAME: "typeClass"})
+    description: str
+    write_permission: str
+    read_permission: str
+    helper_type: List[str] = field(factory=list, metadata={FIELD_NAME: "helperType"})
+    device_models: List[dict] = field(factory=list, metadata={FIELD_NAME: "deviceModels"})
```
Comment (tehAgitto) [inline vmngclient/dataclasses.py:493]: Is it possible to create an object here instead of `dict`?

File: vmngclient/api/template_api.py [hunk 5]
```diff
@@ -531,6 +540,27 @@ def generate_feature_template_payload(
 
         return payload
 
+    def validate_device_model(self, template: FeatureTemplate) -> bool:
+        """Verify if selected template can be used with provided device model"""
+        template_type = self._get_feature_template_types().filter(name=template.type).single_or_default()
+
+        available_devices_for_template = [device["name"] for device in template_type.device_models]
```
Comment (tehAgitto) [inline vmngclient/api/template_api.py:547]: placeholder if `device_models` could be converted into an object.

File: vmngclient/api/template_api.py [hunk 6]
```diff
@@ -531,6 +540,27 @@ def generate_feature_template_payload(
...
+        provided_device_models = [
+            dev_mod.value if type(dev_mod) is DeviceModel else dev_mod for dev_mod in template.device_models
+        ]
```
Comment (tehAgitto) [inline vmngclient/api/template_api.py:550]: placeholder if `device_model` attribute could be converted into only `DeviceModel` object (not `str`)

**Review by tehAgitto — CHANGES_REQUESTED** [review]

**Review by tehAgitto — APPROVED** [review]

File: vmngclient/dataclasses.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 from typing import List, Optional
 
 from attr import define, field  # type: ignore
+from pydantic import BaseModel, Field  # type: ignore
```
Comment (sbasan) [inline vmngclient/dataclasses.py:5]: I think this could be uncommented, recently we have added pydantic to additional dependencies for mypy (pre-commit)

**Review by tomaszwarda — APPROVED** [review]

**Pull request merged**